lib/light-service/action.rb

```diff
@@ -76,6 +93,22 @@ def all_keys
         expected_keys + promised_keys
       end
 
+      def missing_expected_keys(context)
+        expected_keys - context.keys
+      end
+
+      def usable_defaults(context)
+        available_defaults.slice(
+          *(missing_expected_keys(context) & available_defaults.keys)
+        )
+      end
+
+      def extract_default(default, context)
+        return default unless default.respond_to?(:call)
+
+        default.call(context)
+      end
+
```

Comment:
I definitely think it could do with a separate class or module handling key setup as a first step, and then extracting that class into two `expects` and `promises` classes if it needs more refactoring.

Right now, the action module is doing the following things:

1. setting up expects and promises keys (incl. defaults when supplied)
2. executing an action
3. rolling back an action
4. handling callbacks
5. preparing a context
6. verifying key existence before an action

All tasks can be grouped into two categories:

1. Preparation - parsing keys, setting up context, verifying keys
2. Execution - execute, roll back, callbacks

I reckon refactoring the preparation code into a single responsibility will be a huge step forward without necessitating an overhaul. A good effort:result ratio in my opinion.

What do you think. I'd be keen to take a stab at this next month...